### DISCUSSION TITLE
Consulta Distribuição de NFe

### DISCUSSION THREAD

**Discussion by leogregianin** [TadaSoftware/PyNFe#170] [Q&A] [answered]
Dia 25/02/2022 houve a suspensão de 2 dos 3 tipos de serviços da Distribuição de NFe, ou seja, consulta de Notas Fiscal emitidas contra um CNPJ.

Notícia original: Suspensão dos serviços "ConsNSU" e "ConsChNFe" da NT 2014.002. [Fonte: Receita Federal do Brasil](http://www.nfe.fazenda.gov.br/portal/informe.aspx?ehCTG=false&Informe=A6qvFRVbPSA=) 

### **O que afeta o PyNFe?**

Conforme [documentação do wiki](https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribui%C3%A7%C3%A3o) temos 3 tipos diferentes que consultas: 

- Opção 1: distNSU – Distribuição de Conjunto de DF-e a partir do NSU informado
- Opção 2: consNSU – Consulta DF-e vinculado ao NSU informado
- Opção 3: consChNFe – Consulta de NF-e por chave de acesso informada

As opções 2 e 3 foram suspensas. A opção 1 continua em funcionamento.

**Comment by leogregianin** [comment]
Atualização da situação em 04/03/2022.

A Receita Federal publicou a notícia abaixo

https://www.nfe.fazenda.gov.br/portal/informe.aspx?ehCTG=false&Informe=0cu/yBLKrCs=&AspxAutoDetectCookieSupport=1

Vigente a partir de 10/03/2022.

**Comment by devtradeit** [comment, marked as answer]
Uma dúvida, para download dos arquivos não é mais necessário manifestar a ciencia da operação? Pergunto pois rodei o codigo de exemplo que encontrei em [https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribuicao-1](https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribuicao-1) e já saiu baixando tudo. 

Não consegui identificar no codigo se há ou não a manifestação.

Abraço e obrigado.

**Reply by leogregianin** [reply]
@devtradeit, é necessário a [manifestação](https://github.com/TadaSoftware/PyNFe/wiki/Manifesta%C3%A7%C3%A3o-Destinat%C3%A1rio) para baixar o XML completo, essas chaves de acesso que consultou já foram manifestadas, não tem outra opção. Para conferir, pode consultar uma dessas chaves de acesso no [Portal da NFe](https://www.nfe.fazenda.gov.br/portal/principal.aspx) e verificar a data da manifestação.

**Reply by devtradeit** [reply]
@leogregianin tem razão... as notas que foram baixadas realmente ja haviam sido manifestadas anteriormente. No caso de utilizar este mesmo codigo para notas não manifestadas, qual será o resultado? Obrigado pelo retorno!

**Reply by leogregianin** [reply]
@devtradeit se utilizar exatamente o mesmo código [daqui](https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribuicao-1) não irá baixar nenhuma nota não manifestada.

Você tem que encaixar [esse código](https://github.com/TadaSoftware/PyNFe/wiki/Manifesta%C3%A7%C3%A3o-Destinat%C3%A1rio) do envio do evento de manifestação dentro da consulta da distribuição.

A consulta da distribuição pode retornar vários tipos de arquivos, por exemplo, NFe com o XML completo, Eventos como carta de correção ou o Resumo da NFe. Para cada um desses tipos temos um `schema` específico, `XML completo` será `procNFe_v4.00.xsd`, eventos será `procEventoNFe_v1.00.xsd` e o `resumo` será `resNFe_v1.01.xsd`.

Você precisa manifestar todas as chaves de acesso que ainda estão em resumo.

Observe na consulta [aqui](https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribuicao-1) que tem um `if (tipo_schema == 'procNFe_v4.00.xsd'):` para gravar o XML completo em arquivo. Você também terá algo assim (`elif tipo_schema == 'resNFe_v1.01.xsd'):` para chamar o envio do evento de manifestação, e após alguns minutos ou horas, consultar novamente a distribuição e baixar o XML completo dessa chave de acesso manifestada.

**Reply by EduardoFerreira22** [reply]
Tenho uma dúvida se colocar a ciência da operação diretamente  no [Fluxo da Consulta de Distribuicao por NSU](https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU)  para trazer o XML completo, eu já estaria realizando uma das manifestações e não poderia realizar outra nos proximos 180 dias de acordo com esse [artigo](https://intelecta.com.br/blog-do-empresario/diferenca-entre-mde-e-mdfe#:~:text=A%20diferen%C3%A7a%20mais%20b%C3%A1sica%20%C3%A9%20que%20o%20Manifesta%C3%A7%C3%A3o,unidade%20de%20carga%20os%20documentos%20utilizados%20na%20opera%C3%A7%C3%A3o.) .
Nesse caso se preciso criar uma aplicação em que o usuário tem uma interface onde ele poderá realizar a manifestação do documento, ele teria que esperar mais uma hora após esse ato para que a próxima consulta traga o xml completo? @leogregianin @juniortada @flaviosm @z9sb @Sujow

**Comment by flaviosm** [comment]
Pessoal, bom dia!

Rodei o processo de consulta de notas emitidas contra o meu cnpj, e tive o retorno do erro abaixo:

Nova consulta a partir do NSU: 000000000000000
Quantidade de NSUs na consulta atual: 0
cStat: 215
xMotivo: Rejeicao: Falha no esquema xml
maxNSU: 000000000000000
Falha

No parâmetro chave usei o meu cnpj e no CPFCNPJ também! parâmetros NSU, ultNSU, maxNSU e cStat deixei o padrão igual a 0

**Reply by leogregianin** [reply]
> Pessoal, bom dia!
> 
> Rodei o processo de consulta de notas emitidas contra o meu cnpj, e tive o retorno do erro abaixo:
> 
> Nova consulta a partir do NSU: 000000000000000 Quantidade de NSUs na consulta atual: 0 cStat: 215 xMotivo: Rejeicao: Falha no esquema xml maxNSU: 000000000000000 Falha
> 
> No parâmetro chave usei o meu cnpj e no CPFCNPJ também! parâmetros NSU, ultNSU, maxNSU e cStat deixei o padrão igual a 0

Você está usando a consulta [com a opção 1]((https://github.com/TadaSoftware/PyNFe/wiki/Consulta-Distribui%C3%A7%C3%A3o)) e o campo `CHAVE` deve ficar em branco.

**Comment by z9sb** [comment]
Olá, tudo bem?
Estou tentando fazer alguns downloads, ele até me retorna o zip_resposta, mas não consigo encontrar o arquivo que e descompactado.
Creio que não esteja detectando notas para o download das mesmas, sendo que já foram manifestadas.
Saberia me informar se devo fazer algo mais @leogregianin?

**Reply by leogregianin** [reply]
Você precisa, explicitamente, gravar o arquivo. 
Veja [esse fluxo](https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao) e altere a pasta de gravação do XML.

**Reply by ibemoreira** [reply]
Segui esse fluxo e gerou em meu diretório varios xmls "consulta_distrib_gzip-000000000185558.xml" Pelo que eu entendi é uma lista de varis documentos compactados, minha pergunta é como faço para baixar o xmls desses documentos que estão nessa lista? 

Imagino que após seguir esse fluxo tenho que descompactar o resumo que esta no xml que gerou pegar a chave manifestar e rodar o fluxo novamente que vai baixar o XML da NF seria isso?

**Reply by leogregianin** [reply]
Exatamente isso @ibemoreira.

**Reply by ibemoreira** [reply]
Legal! agora vou atrás de conseguir ler esse </docZip> para pegar a chave.

**Reply by ibemoreira** [reply]
@leogregianin  Desculpa incomodar pode parecer simples mas sou um iniciante em python. 
 
Dentro do xml eu tenho:
```
 <docZip NSU="000000000185509" schema="procEventoNFe_v1.00.xsd">H4sI............</docZip>
<docZip NSU="000000000185510"
```
Acredito que tenho que descompactar essa tag ou decodificar para pegar o ID da NF e manifesta-la, porém não consegui ler o conteúdo dessa tag docZip.

**Reply by leogregianin** [reply]
O XML de retorno pode conter 3 tipos:
 - a nota fiscal completa, que a variável `tipo_schema` será `procNFe_v4.00.xsd`
 - um evento qualquer, que será `procEventoNFe_v1.00.xsd`
 - e o resumo da nota para manifestar, que será `resNFe_v1.01.xsd`

Eu atualizei o wiki com o exemplo (https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao)

**Reply by ibemoreira** [reply]
Top brother vou testar aqui! Obrigado pela ajuda!

**Reply by ibemoreira** [reply]
@leogregianin  Estou me sentido chato mas com esse novo codigo esta dando um erro nessa linha 
`zip_resposta = resposta.xpath('//ns:retDistDFeInt/ns:loteDistDFeInt/ns:docZip', namespaces=self.ns)[contador_xml].text`

Entendi como se o "self" não tivesse declarado ou algo do tipo.

**Reply by leogregianin** [reply]
Não tem o `self` alí. Atualizei o wiki

**Reply by ibemoreira** [reply]
Boa tarde, passei a tarde tentando resolver por conta própria mas esse código esta retornando o seguinte erro na linha 70 
`typeError: Type str cannot be serialized`

Desde já meu muito obrigado @leogregianin

**Reply by ibemoreira** [reply]
Consegui resolver esse erro mudando o código para isso :
`                chave_acesso_nfe = resposta_descompactado.xpath('//ns:resNFe/ns:chNFe', namespaces=ns)[0].text
                # Gerar ciência da operação
                print(f'Chave: {chave_acesso_nfe}')`

 E no codigo do wiki esta assim: 
`# Ler o resumo
                xml = etree.tostring(xml_completo).decode('utf-8')
                chave_acesso_nfe = xml.xpath('//ns:resNFe/ns:chNFe', namespaces=ns)[0].text
                # Gerar ciência da operação`

Removi a linha 70 e no xpath invés de chamar o "xml" chamei o "resposta_descompactado"

Feliz por conseguir resolver sozinho kkkk

**Comment by Sujow** [comment]
Utilizando o código padrão disponibilizado [aqui]( https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU), e claro, alterando os dados para os meus, eu recebo o seguinte retorno:

Nova consulta a partir do NSU: <Response [200]>
Quantidade de NSUs na consulta atual: 0
cStat: 656
xMotivo: Rejeicao: Consumo Indevido (Deve ser utilizado o ultNSU nas solicitacoes subsequentes. Tente apos 1 hora)
maxNSU: 000000000000000
Falha

Sempre esse mesmo retorno.

**Reply by leogregianin** [reply]
> Utilizando o código padrão disponibilizado [aqui](https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU), e claro, alterando os dados para os meus, eu recebo o seguinte retorno:
> 
> Nova consulta a partir do NSU: <Response [200]> Quantidade de NSUs na consulta atual: 0 cStat: 656 xMotivo: Rejeicao: Consumo Indevido (Deve ser utilizado o ultNSU nas solicitacoes subsequentes. Tente apos 1 hora) maxNSU: 000000000000000 Falha
> 
> Sempre esse mesmo retorno.

@Sujow esse é o comportamento esperado pelo webservice.

Sugiro ler as mensagens anteriores desse mesmo tópico e essa notícia https://www.nfe.fazenda.gov.br/portal/informe.aspx?ehCTG=false&Informe=0cu/yBLKrCs=&AspxAutoDetectCookieSupport=1

**Reply by Sujow** [reply]
Entendi, Leo!
Testei com outras empresas e obtive outro retorno, esse outro retorno exibia a quantidade e tudo mais, só não baixou os xml.

O que devo fazer para que o download ocorra? 
Em qual pasta será salvo?

**Reply by juniortada** [reply]
https://github.com/TadaSoftware/PyNFe/wiki/Fluxo-da-Consulta-de-Distribuicao-por-NSU
